src/core/server/index.ts

```diff
@@ -665,3 +649,42 @@ export const config = {
     appenders: appendersSchema as Type<AppenderConfigType>,
   },
 };
+
+/**
+ * Public version of RequestHandler, default-scoped to {@link RequestHandlerContext}
+ * See [@link RequestHandler}
+ * @public
+ */
+type PublicRequestHandler<
+  P = unknown,
+  Q = unknown,
+  B = unknown,
+  Context extends RequestHandlerContext = RequestHandlerContext,
+  Method extends RouteMethod = any,
+  ResponseFactory extends KibanaResponseFactory = KibanaResponseFactory
+> = RequestHandler<P, Q, B, Context, Method, ResponseFactory>;
+
+export type { PublicRequestHandler as RequestHandler, RequestHandler as BaseRequestHandler };
+
+/**
+ * Public version of IRouter, default-scoped to {@link RequestHandlerContext}
+ * See [@link IRouter}
+ * @public
+ */
+type PublicRouter<ContextType extends RequestHandlerContext = RequestHandlerContext> =
+  IRouter<ContextType>;
+
+export type { PublicRouter as IRouter, IRouter as IBaseRouter };
```

Comment:
This is the part I'm less happy about. If access to router via `coreSetup.http.createRouter` are properly scoped to use `RequestHandlerContext`, there were still a lot of places where API consumers are importing and using the unscoped types imported directly from core, e.g

```ts
import { IRouter } from `core`

const defineRoute = (router: IRouter) => {
   router.get(config, (ctx, req, res) => {
      // ctx is of type `RequestHandlerContextBase` here
   )}
}
```

So I needed to have the public version of some of our types (namely `IRouter, RequestHandler and HttpServiceSetup`)

Introducing new types wasn't really an option here given the amount of changes this would have required in the whole codebase (likely 2k+ file changes we're talking about here), so I made the pragmatic decision to export pre-scoped version of these types under the current name. I'm also forced to export the original versions, renamed too (for generated docs, mostly), as the public renamed types depends on the original types. 

After some long thinking, I believe the best approach would be to keep the current name for the public types, and rename the internal types. E.g the internal `IRouter` could be renamed `InternalRouter` or something similar). However, even if that means less changes than changing the public types, the usages of these types within core is still quite significant, so I didn't want to do it in this PR.

If we're fine with the approach, I'll open a follow-up issue to discuss our options.

Reply:
> I believe the best approach would be to keep the current name for the public types, and rename the internal types

++ The fewer changes we have to public types, the better. The changes we're making are already significant, IMO, the less risky option for core consumers is to minimize the effects of these changes. While I agree that there's still a significant risk we get this wrong within `core`, as long as the transition period is small enough, hopefully, it'll be ok. 

Do you think it's worth documenting the risks in the code now?

Reply:
> Do you think it's worth documenting the risks in the code now?

I think it's better to add a follow-up issue and link it in https://github.com/elastic/kibana/issues/134112, I will do that before the merge

Reply:
Created https://github.com/elastic/kibana/issues/135773